From foo@baz Tue May  1 16:18:19 PDT 2018
From: Eric Dumazet <edumazet@google.com>
Date: Fri, 23 Feb 2018 08:12:42 -0800
Subject: percpu: add a schedule point in pcpu_balance_workfn()

From: Eric Dumazet <edumazet@google.com>

[ Upstream commit accd4f36a7d11c2d54544007eb65e10604dcf2f5 ]

When a large BPF percpu map is destroyed, I have seen
pcpu_balance_workfn() holding cpu for hundreds of milliseconds.

On KASAN config and 112 hyperthreads, average time to destroy a chunk
is ~4 ms.

[ 2489.841376] destroy chunk 1 in 4148689 ns
...
[ 2490.093428] destroy chunk 32 in 4072718 ns

Signed-off-by: Eric Dumazet <edumazet@google.com>
Signed-off-by: Tejun Heo <tj@kernel.org>
Signed-off-by: Sasha Levin <alexander.levin@microsoft.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---
 mm/percpu.c |    1 +
 1 file changed, 1 insertion(+)

--- a/mm/percpu.c
+++ b/mm/percpu.c
@@ -1611,6 +1611,7 @@ static void pcpu_balance_workfn(struct w
 			spin_unlock_irq(&pcpu_lock);
 		}
 		pcpu_destroy_chunk(chunk);
+		cond_resched();
 	}
 
 	/*
